src/rust/src/address_ffi.rs

```diff
@@ -0,0 +1,219 @@
+    unsafe extern "C" fn(ua: Option<UnifiedAddressObj>, index: usize) -> usize;
+pub type GetReceiverDataCb =
+    unsafe extern "C" fn(ua: Option<UnifiedAddressObj>, index: usize, data: *mut u8);
+
+fn network_from_cstr(network: *const c_char) -> Option<Network> {
+    match unsafe { CStr::from_ptr(network) }.to_str().unwrap() {
+        "main" => Some(Network::Main),
+        "test" => Some(Network::Test),
+        "regtest" => Some(Network::Regtest),
+        s => {
+            tracing::error!("Unsupported network type string '{}'", s);
+            None
+        }
+    }
+}
+
+struct UnifiedAddressHelper {
+    net: Network,
+    ua: unified::Address,
+}
+
+impl FromAddress for UnifiedAddressHelper {
+    fn from_unified(
+        net: Network,
+        ua: unified::Address,
+    ) -> Result<Self, zcash_address::UnsupportedAddress> {
+        Ok(Self { net, ua })
+    }
+}
+
+impl UnifiedAddressHelper {
+    fn to_cpp(
+        self,
+        network: Network,
+        ua_obj: Option<UnifiedAddressObj>,
+        sapling_cb: Option<AddReceiverCb>,
+        p2sh_cb: Option<AddReceiverCb>,
+        p2pkh_cb: Option<AddReceiverCb>,
+        unknown_cb: Option<UnknownReceiverCb>,
+    ) -> bool {
+        if self.net != network {
+            tracing::error!(
+                "Unified Address is for {:?} but node is on {:?}",
+                self.net,
+                network
+            );
+            return false;
+        }
+
+        self.ua
+            .receivers()
+            .into_iter()
+            .map(|receiver| match receiver {
+                unified::Receiver::Orchard(data) => {
+                    // ZIP 316: Senders MUST reject Unified Addresses in which any
+                    // constituent address does not meet the validation requirements of
+                    // its Receiver Encoding.
+                    // TODO: Add this API to the orchard crate.
+                    // if let Err(e) = orchard::Address::from_bytes(data) {
+                    //     tracing::error!("{}", e);
+                    //     false
+                    // } else {
+                    {
+                        unsafe {
+                            // TODO: Replace with Orchard support.
+                            (unknown_cb.unwrap())(ua_obj, 0x03, data.as_ptr(), data.len())
+                        }
+                    }
+                }
+                unified::Receiver::Sapling(data) => {
+                    // ZIP 316: Senders MUST reject Unified Addresses in which any
+                    // constituent address does not meet the validation requirements of
+                    // its Receiver Encoding.
+                    if sapling::PaymentAddress::from_bytes(&data).is_none() {
+                        tracing::error!("Unified Address contains invalid Sapling receiver");
+                        false
+                    } else {
+                        unsafe { (sapling_cb.unwrap())(ua_obj, data.as_ptr()) }
```

Comment:
We cannot avoid a `std::vector` on the C++ side due to unknown typecodes, so we would need to deal with this regardless. But that could be a nice refactor for handling known typecodes.

Reply:
I don't think it is a bad encoding; in a sense it's more typesafe, because it expresses that you can only have at most one of each receiver type. You would probably still want a way to iterate over the receivers, though.